scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -418,6 +430,96 @@ int BPF_PROG(tp_task_rename, struct task_struct *p, const char *buf)
 	return 0;
 }
 
+static void refresh_overrides(int nr_layers)
+{
+	int i, layer_idx;
+	u64 override_gen;
+	struct task_ctx *tctx;
+	struct override_ctx *octx;
+	struct task_struct *p;
+
+	bpf_for(layer_idx, 0, nr_layers) {
+		struct layer *layer = &layers[layer_idx];
+		override_gen = override_gens[layer_idx];
+		if (!(octx = lookup_override_ctx(layer_idx)))
+			continue;
+		if (octx->gen == override_gen)
+			continue;
+
+		// If we had proper locks with userspace then this may be an
+		// allowed state for wrapping the override_gen. For now we'll
+		// allow resetting the gen to 0.
+		if (octx->gen < override_gen && octx->gen != 0) {
+			scx_bpf_error("override generation is lower than set generation");
+			return;
+		}
+
+		override_gens[layer_idx] = octx->gen;
+
+		bpf_for(i, 0, octx->nr_overrides) {
+			if (i > MAX_LAYER_OVERRIDES) {
+				scx_bpf_error("too many overrides");
+				return;
+			}
+
+			p = bpf_task_from_pid(octx->task_overrides[i]);
+			if (!p)
+				continue;
+
+			if (!(tctx = lookup_task_ctx(p))) {
+				bpf_task_release(p);
+				continue;
+			}
+			if (octx->gen < tctx->override_gen && octx->gen != 0) {
+				bpf_task_release(p);
+				scx_bpf_error("task override generation is invalid");
+				return;
+			}
+			if (octx->gen == tctx->override_gen) {
+				bpf_task_release(p);
+				continue;
+			}
+			if (tctx->layer >= 0 && tctx->layer < nr_layers)
+				__sync_fetch_and_add(&layers[tctx->layer].nr_tasks, -1);
+			/*
+			 * XXX: Need to adjust the vtime delta with the
+			 * previous layer similar to maybe_refresh_layer.
+			 * However, from this context the value is read only so
+			 * it may be incorrect.
+			 *
+			 * p->scx.dsq_vtime = layer->vtime_now;
+			 *
+			 */
+			tctx->layer = layer_idx;
```

Comment:
So, the way to do this, I think, is just updating the generation numbers here and what the target index should be and let `maybe_refresh_layer()` do the actual update. That way, it'll update in a synchronized manner and vtime will be taken care of too.

Reply:
Yeah, I think that makes sense... keep the updates in one place.